File: converter/java/src/main/java/com/mlt/vector/fsstdictionary/StringFsstDictionaryVector.java
package com.mlt.vector.fsstdictionary;

import com.mlt.vector.BitVector;
import com.mlt.vector.VariableSizeVector;

import java.nio.ByteBuffer;
import java.nio.IntBuffer;

public class StringFsstDictionaryVector extends VariableSizeVector<String> {
    //TODO: extend from StringVector

    private IntBuffer offsets;
    private int[] offsetBuffer;
    private IntBuffer indexBuffer;

    public StringFsstDictionaryVector(String name, IntBuffer indexBuffer, IntBuffer lengthBuffer, ByteBuffer dictionaryBuffer,
                                     IntBuffer symbolLengthBuffer,  ByteBuffer symbolTableBuffer) {
        super(name, lengthBuffer, dictionaryBuffer);
        setBuffer(indexBuffer);
    }

    public StringFsstDictionaryVector(String name, BitVector nullabilityBuffer, IntBuffer indexBuffer, IntBuffer lengthBuffer,
                                  ByteBuffer dictionaryBuffer, IntBuffer symbolLengthBuffer,  ByteBuffer symbolTableBuffer) {
        super(name, nullabilityBuffer, lengthBuffer, dictionaryBuffer);
        setBuffer(indexBuffer);
    }

    private void setBuffer(IntBuffer indexBuffer){
        this.indexBuffer = indexBuffer;
        this.offsetBuffer = new int[indexBuffer.capacity()];
    }

    @Override
    protected String getValueFromBuffer(int index) {
        return "";
    }
}

File: ts/src/vector/fsst-dictionary/stringFsstDictionaryVector.ts
import { VariableSizeVector } from "../variableSizeVector";
import type BitVector from "../flat/bitVector";
import { decodeFsst } from "../../decoding/fsstDecoder";
import { decodeString } from "../../decoding/decodingUtils";

/** Mutable cache shared by the FSST child columns of one SharedDict. */
export type FsstDictionaryCache = {
    /** Undefined until one member of the SharedDict group decodes the dictionary on first access. */
    decodedDictionary?: Uint8Array;
};

export class StringFsstDictionaryVector extends VariableSizeVector<Uint8Array, string> {
    // TODO: extend from StringVector
    private symbolLengthBuffer: Uint32Array;
    private decodedDictionary: Uint8Array;

    constructor(
        name: string,
        private readonly indexBuffer: Uint32Array,
        offsetBuffer: Uint32Array,
        dictionaryBuffer: Uint8Array,
        private readonly symbolOffsetBuffer: Uint32Array,
        private readonly symbolTableBuffer: Uint8Array,
        nullabilityBuffer: BitVector,
        /** Cache shared by the FSST child columns of one SharedDict. */
        private readonly sharedDictionaryCache?: FsstDictionaryCache,
    ) {
        super(name, offsetBuffer, dictionaryBuffer, nullabilityBuffer ?? indexBuffer.length);
    }

    protected getValueFromBuffer(index: number): string {
        if (this.decodedDictionary == null) {
            this.decodedDictionary = this.sharedDictionaryCache?.decodedDictionary;
            if (this.decodedDictionary == null) {
                this.decodedDictionary = this.decodeDictionary();
                if (this.sharedDictionaryCache) {
                    this.sharedDictionaryCache.decodedDictionary = this.decodedDictionary;
                }
            }
        }

        const offset = this.indexBuffer[index];
        const start = this.offsetBuffer[offset];
        const end = this.offsetBuffer[offset + 1];
        return decodeString(this.decodedDictionary, start, end);
    }

    private decodeDictionary(): Uint8Array {
        if (this.symbolLengthBuffer == null) {
            this.symbolLengthBuffer = this.offsetToLengthBuffer(this.symbolOffsetBuffer);
        }
        return decodeFsst(this.symbolTableBuffer, this.symbolLengthBuffer, this.dataBuffer);
    }

    // TODO: get rid of that conversion
    private offsetToLengthBuffer(offsetBuffer: Uint32Array): Uint32Array {
        const lengthBuffer = new Uint32Array(offsetBuffer.length - 1);
        let previousOffset = offsetBuffer[0];
        for (let i = 1; i < offsetBuffer.length; i++) {
            const offset = offsetBuffer[i];
            lengthBuffer[i - 1] = offset - previousOffset;
            previousOffset = offset;
        }

        return lengthBuffer;
    }
}
